articles/055_ros_parameter_design.md

```diff
@@ -109,17 +109,28 @@ Each parameter consists of a key and a value.
 The key is a string value.
 The value can be one of the following datatypes:
 
-- `float64`
+- `bool`
 - `int64`
+- `float64`
 - `string`
-- `bool`
-- `bytes[]`
+- `byte[]`
+- `bool[]`
+- `int64[]`
+- `float64[]`
+- `string[]`
 
-The datatypes are chosen as non-complex datatypes, as defined in the [interface definitions article](articles/interface_definition.html)
+The datatypes are chosen as non-complex datatypes, as defined in the [interface definitions article](interface_definition.html)
 The full complement of datatypes of different bitdepth and unsigned types are avoided to allow interpretation from text based configuration files.
 
-`bytes` are included to allow the storage of binary blobs.
-It's use is not recommended but can be very convenient, and explicitly supporting it is better than having people try to abuse other datatypes such as strings.
+Array data types support a variety of use cases, e.g.,
+- `byte[]` is included to allow the storage of binary blobs.
+Its use is not recommended but can be very convenient, and explicitly supporting it is better than having people try to abuse other datatypes such as strings.
+- `bool[]` can be useful as a mask for other array parameters.
+- `int64[]` and `float64[]` can be used to express numerical vector and matrix parameters.
+- `string[]` can be used to express groups of names, such as a set of robots in a multi-robot context.
+
+While the use of array parameters increases the complexity of the API, their omission would necessitate complex naming schemes for parameters like matrices, which are common in robotics.
+They should not be abused, however; users should rely on namespacing and explicit variable names wherever possible.
```

Comment:
Should it still be possible to query single elements from array parameters?

let's say in yaml:

```
robot:
  joints:
    - joint1
    - joint2
    - joint3
```
Should we have one parameter `robot/joints` which will return the three values? If so, how would nested arrays look like?
```
joint_map:
  -
    - joint1
    - joint2
  -
    - my_joint1
    - my_joint2
```
I am not saying we necessarily support this - you could get around all this by using maps instead of nested arrays - but it's maybe worth mentioning it.

Reply:
I don't actually feel strongly one way or another about this. Anyone else?

Reply:
Actually @Karsten1987 if you wanted to do that I would recommend using new keys:

```yaml
joint_map:
  keys:
    - joint1
    - joint2
  values:
    - my_joint1
    - my_joint2
```

Which would become parameters like `joint_map/keys` -> `[joint1, joint2]`. Or you could do something like this:

```yaml
joint_map:
  joint1: my_joint1
  joint2: my_joint2
```

Which would become parameters like `joint_map/joint1` -> `my_joint1`, etc...

----

I think a more general use for this would be a matrix, e.g. a 3 by 3:

```yaml
my_matrix: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
```

Essentially should there be lists of lists.

I feel like this is a critical feature, but I could be convinced otherwise I think.

Reply:
@tfoote I actually think my first yaml example should be rendered as one parameter with a collection of three values. https://symfony.com/doc/current/components/yaml/yaml_format.html#collections

There is no direct key-value pair, but rather a list of three values. So the parameter I'd expect would be `robot/joints` --> [`joint1`, `joint2`, `joint3`]

@wjwwood I also would recommend using maps/key-value pairs instead, I solely wanted to point out that yaml supports such a syntax and we should mention how we would render it.

Reply:
This first example you gave is the expected way to represent lists of parameters as yaml, why do you think it needs to handled specially. 

We do need to handle yaml with non-homogeneous lists. Our current proposal does not handle that I think.

Reply:
Also why did you @tfoote? I'm confused.

Reply:
Hmm, must be something weird with github. I also see your comment twice here in that thread.

Reply:
I didn't see @tfoote's comment until just now... weird I explicitly looked for it before...

Reply:
So for ROS 1, I've been interpreting arrays as matrices by just reshaping them. It puts the onus on the package authors to interpret the parameters correctly, but I don't personally see an issue with that. How much complexity are we adding by supporting lists of lists? Would we support arbitrary list depths (e.g., lists of lists of lists of lists)?

Reply:
It looks like ROS 1 only supports one dimensional lists as well as only supporting homogeneous lists (can someone confirm that in this document? with links to docs or code if possible).

That would be a hint that maybe supporting the same set of features is sufficient for ROS 2 as well. I'm interested to see what others think.

If we only support homogenous lists, then supporting indefinite dimensionality to the lists seems like a limited amount of overhead, but I could be wrong about that. It might require some prototyping and brainstorming to find the corner cases that bite.